tomb: recolor sarcophagus config

rdutta opened the pull request:
#39 Adds configs for recoloring the reward chest/sarcophagus walls.

LlemonDuck commented:
Thanks for the submission, I'd love to add this. Do you have a video confirming it works in a raid?

rdutta commented:
https://user-images.githubusercontent.com/38548565/212801608-9c0a5c16-cbc3-4d7e-9040-b10582ac3348.mp4

I have tested the purple recolor options as well, but only in a solo raid.

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/TombsOfAmascutConfig.java:
```diff
@@ -394,6 +396,62 @@ default int chestAudioVolume()
+	String SECTION_REWARD_SARCOPHAGUS = "rewardSarcophagusSection";
+
+	@ConfigItem(
+		name = "Recolor white",
```
Config names should be Title Case

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/TombsOfAmascutConfig.java:
```diff
@@ -394,6 +396,62 @@ default int chestAudioVolume()
+	}
+
+	@ConfigItem(
+		name = "Color",
```
Regular Loot Colour

note british spelling

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/TombsOfAmascutConfig.java:
```diff
@@ -394,6 +396,62 @@ default int chestAudioVolume()
+		return Color.BLACK;
+	}
+
+	@ConfigItem(
+		name = "Recolor purple",
+		description = "Recolor the sarcophagus purple lighting.",
+		position = 703,
+		keyName = "sarcophagusRecolorPurple",
+		section = SECTION_REWARD_SARCOPHAGUS
+	)
+	default RecolorMode sarcophagusRecolorPurple()
+	{
+		return RecolorMode.OFF;
+	}
```
Rather than a 3-way enable, you should allow players to rename personal purples and others' purples separate colours.

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/TombsOfAmascutConfig.java:
```diff
@@ -394,6 +396,62 @@ default int chestAudioVolume()
+		keyName = "sarcophagusPurpleRecolor",
+		section = SECTION_REWARD_SARCOPHAGUS
+	)
+	default Color sarcophagusPurpleRecolor()
+	{
+		return Color.BLACK;
+	}
```
If you extract the default model colours, you can remove the enable config options to just have the default values be the in-game colour instead. Would help to reduce the config density, but requires getting another purple to find that value.

LlemonDuck replied:
This one could also be done in a later update, doesn't need to be done for now.

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+public class RewardSarcophagus implements PluginLifecycleComponent
+{
+
+	private static final int[] VARBIT_MULTILOC_IDS_CHEST = new int[]{
+		14356, 14357, 14358, 14359, 14360, 14370, 14371, 14372
+	};
```
Set<Integer> will let you replace the for-loop finder in `parseVarbits` with a Set#contains. Reads a little nicer.

rdutta replied:
confused on this one

LlemonDuck replied:
Ah, I see I misread that one. It's good as-is.

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+
+		final String key = event.getKey();
+
+		if (key.startsWith("sarcophagus"))
```
I'd rather have key name constants in the config file and a `static Set<String>` in this class which references them, and finally a Set#contains here.

Also fine with you placing a `String KEY_PREFIX_LOOT_RECOLOR = "sarcophagusRecolour"` and using `keyName = KEY_PREFIX_LOOT_RECOLOR + "White"`, then checking for startsWith that prefix as well.

At least some sort of reference-based comparison here in case the config key names change.

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+	@Subscribe
+	public void onConfigChanged(final ConfigChanged event)
+	{
+		if (!event.getGroup().equals(TombsOfAmascutConfig.CONFIG_GROUP) || wallObjects.isEmpty()) return;
```
expand with braces

it's not in the checkstyle but it is part of the [RuneLite Code Conventions](https://github.com/runelite/runelite/wiki/Code-Conventions) which the plugin uses (they don't explicitly mention it there either but it's in the xml config provided).

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+ * BSD 2-Clause License
+ * <p>
+ * Copyright (c) 2023, rdutta
+ * Copyright (c) 2020, AnkouOSRS
```
Can you provide me with a link of where this attribution comes from so I can confirm the license?

rdutta replied:
https://github.com/AnkouOSRS/cox-light-colors/blob/master/LICENSE
https://github.com/AnkouOSRS/cox-light-colors/blob/master/src/main/java/com/coxlightcolors/CoxLightColorsPlugin.java

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+	{
+		final WallObject wallObject = event.getWallObject();
+
+		if (wallObject.getId() != WALL_OBJECT_ID_SARCOPHAGUS) return;
```
expand

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+
+		if (canCloneDefaultFaceColors())
+		{
+			final Model model = wallObject.getRenderable1().getModel();
```
This model can be null

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+	{
+		clientThread.invokeLater(() ->
+		{
+			sarcophagusIsPurple = client.getVarbitValue(VARBIT_ID_SARCOPHAGUS) % 2 != 0;
```
For my own understanding, the varbit is 0 (others') or 2 (mine) when purple, and 1 when white? Or are those values different?

I'm seeing on the [wiki's varbs site](https://chisel.weirdgloop.org/varbs/display?varbit=14373) that the values range from 0-3. Do you know what each is for?

rdutta replied:
No idea, I pulled that info from the dev channel https://discord.com/channels/301497432909414422/419891709883973642/1012999980954435614

If that is the case then checking for the multilocs can be removed I suppose. I don't have access to groups to raid with to test it out. I just know the "spinning key" multiloc object is not present if the purple is yours in a raid (based off youtube videos).

LlemonDuck replied:
I've reached out to the wiki team for some more information to see if they know what the values indicate.

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+		});
+	}
+
+	private boolean canCloneDefaultFaceColors()
```
If I understand correctly, this is just checking whether we already have the colours, which the method name doesn't indicate. `needDefaultFaceColours` or `missingDefaultFaceColours` would be more descriptive.

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+		}
+	}
+
+	private void recolor(final WallObject wallObject, final Color color)
```
Move upward so that it's beside the other method called recolor

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+		return (encode_hue << 10) + (encode_saturation << 7) + (encode_brightness);
+	}
+
+	private static void copyArrays(final int[] src1, final int[] dest1, final int[] src2,
```
i'd rather this is just `copyArray` singular called 3x

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+			return;
+		}
+
+		System.arraycopy(src1, 0, dest1, 0, dest1.length);
+		System.arraycopy(src2, 0, dest2, 0, dest2.length);
+		System.arraycopy(src3, 0, dest3, 0, dest3.length);
```
should check for errors and attempt to reset the colours to their previous values if anything goes wrong

e.g. if copy 1 succeeds but copy 2 fails, we are in an inconsistent state

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,277 @@
+		{
+			sarcophagusIsPurple = client.getVarbitValue(VARBIT_ID_SARCOPHAGUS) % 2 != 0;
+
+			for (final int varbitId : VARBIT_MULTILOC_IDS_CHEST)
```
set `purpleIsMine = true` before this loop

LlemonDuck reviewed (changes requested):
Thanks, most of the comments I've given are stylistic and should be easy enough to remedy.

Would also be nice to get this tested in a group to ensure the purpleIsMine check succeeds. Depending on order of events, it may not be set correctly before the wall objects spawn. I'll use this branch for my future runs in case I get a purple and can test it as well.

LlemonDuck commented on src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java:
```diff
@@ -0,0 +1,293 @@
+import net.runelite.client.events.ConfigChanged;
+
+@Singleton
+@RequiredArgsConstructor(onConstructor_ = @Inject)
```
Replace this with `@Inject` on applicable fields

rdutta commented:
I found some more information on the face colors. Only the faceColors1 array needs changed when recoloring.

The default arrays for white and purple
```java
	private static final int[] FACE_COLORS1_WHITE = new int[60];
	private static final int[] FACE_COLORS1_PURPLE = new int[60];

	static
	{
		Arrays.setAll(FACE_COLORS1_WHITE, i -> i < 30 ? 126 : 110);
		Arrays.fill(FACE_COLORS1_PURPLE, 50142);
	}
```

I am wondering how to convert the purple value (hsb? hsl?), 50142, to rgb to put in the config as a default value.

LlemonDuck commented:
> I found some more information on the face colors. Only the faceColors1 array needs changed when recoloring.
> 
> The default arrays for white and purple
> 
> ```java
> 	private static final int[] FACE_COLORS1_WHITE = new int[60];
> 	private static final int[] FACE_COLORS1_PURPLE = new int[60];
> 
> 	static
> 	{
> 		Arrays.setAll(FACE_COLORS1_WHITE, i -> i < 30 ? 126 : 110);
> 		Arrays.fill(FACE_COLORS1_PURPLE, 50142);
> 	}
> ```
> 
> I am wondering how to convert the purple value (hsb? hsl?), 50142, to rgb to put in the config as a default value. edit: seems to be `#30075E`

1. Can you try using `JagexColor.rgbToHsl` instead of the one copied from cox lightbeams and let me know if that still works + looks the same?
2. `JagexColor` also has utilities for unpacking an HSL into its components, which can then be placed into any online converter ((like this one)[https://www.rapidtables.com/convert/color/hsl-to-rgb.html])

LlemonDuck commented:
Now that I think about it, this (color extraction work) all may not be worth the effort and might even confuse some users who don't know how to reset it. If you'd like to pursue it and can get it to work, it could be nice, but if you'd rather also not and go back to the checkboxes, I am okay with that too.

rdutta commented:
I have reverted the config options and how the default face colors are initialized. It still needs retesting with a purple chest, mainly to determine if the varbits are set before the region is loaded and the module is enabled (from my previous testing it is). Additionally I assume that the plugin's startUp is run on the client thread so I removed the invokeLater.

LlemonDuck commented:
Thanks, I'll try and test this out as well and read through the code a bit later. One note though, startUp is *not* guaranteed to run on client thread so you should re-add the clientThread#invokeLater.

LlemonDuck commented:
Everything looks good to me, do you have any outstanding things to wrap up or is this ready to merge?

LlemonDuck commented:
Now that we have the separate config toggle, we can provide default colours. I'm a big fan of pastels, what do you think of these choices?

white: gold to match of the rest of the vault room
![image](https://user-images.githubusercontent.com/1868974/214254978-8739e6f6-58bb-42b3-b4f0-159056718310.png)

self: pink which pulls toward the red side of the default purple
![image](https://user-images.githubusercontent.com/1868974/214255159-79651580-b717-4bfa-860d-c0490b279b03.png)

other: blue which pulls toward the blue side of the default purple
![image](https://user-images.githubusercontent.com/1868974/214255379-57ba19a3-8538-4b78-abfe-4b1f82f6d76e.png)

rdutta commented:
LGTM
